File: Sorting Algorithms/Heap Sort/HeapSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void maxHeapify(int* heap,int i,int size)
{
	int l = 2 * i + 1;
	int r = 2 * i + 2;
	int largest;
	if (l < size && heap[l] > heap[i])
		largest = l;
	else
		largest = i;
	if (r < size && heap[r] > heap[largest])
		largest = r;
	if (largest != i)
	{
		swap(heap[i], heap[largest]);
		maxHeapify(heap,largest,size);
	}
}
void buildHeap(int* heap,int size)
{
	for (int i = size / 2 - 1; i >= 0; i--)
		maxHeapify(heap,i,size);
}
void heapSort(int* heap,int size)
{
	buildHeap(heap,size);
	for (int i = size - 1; i > 0; i--)
	{
		swap(heap[0], heap[i]);
		maxHeapify(heap,0,i);
	}
}
int main()
{
	int heap[11] = { 22,1,18,6,5,0,32,16,11,13,7 };
	heapSort(heap,11);
	for (int i = 0; i < 11; i++)
		cout << heap[i] << endl;
	return 0;
}

File: Sorting Algorithms/Heap Sort/HeapSort.py
def maxHeapify(heap, i, size):
    l = 2 * i + 1
    r = 2 * i + 2
    if l < size and heap[l] > heap[i]:
        largest = l
    else:
        largest = i
    if r < size and heap[r] > heap[largest]:
        largest = r
    if largest != i:
        heap[i], heap[largest] = heap[largest], heap[i]
        maxHeapify(heap, largest, size)

def buildHeap(heap, size):
    for i in range(int(size/2 - 1), -1, -1):
        maxHeapify(heap, i, size)

def heapSort(heap, size):
    buildHeap(heap, size)
    for i in range(size - 1, 0, -1):
        heap[0], heap[i] = heap[i], heap[0]
        maxHeapify(heap, 0, i)

heap = [22,1,18,6,5,0,32,16,11,13,7]
heapSort(heap, len(heap))
print(heap)